neo/IO/Caching/DataCache.cs

```diff
@@ -118,16 +118,58 @@ public void DeleteWhere(Func<TKey, TValue, bool> predicate)
             }
         }
 
+        /// <summary>
+        /// Find the entries that start with the `key_prefix`
+        /// </summary>
+        /// <param name="key_prefix">Must maintain the deserialized format of TKey</param>
+        /// <returns>Entries found with the desired prefix</returns>
         public IEnumerable<KeyValuePair<TKey, TValue>> Find(byte[] key_prefix = null)
         {
+            IEnumerable<(byte[], TKey, TValue)> cached;
             lock (dictionary)
             {
-                foreach (var pair in FindInternal(key_prefix ?? new byte[0]))
-                    if (!dictionary.ContainsKey(pair.Key))
-                        yield return pair;
-                foreach (var pair in dictionary)
-                    if (pair.Value.State != TrackState.Deleted && (key_prefix == null || pair.Key.ToArray().Take(key_prefix.Length).SequenceEqual(key_prefix)))
-                        yield return new KeyValuePair<TKey, TValue>(pair.Key, pair.Value.Item);
+                cached = dictionary
+                    .Where(p => p.Value.State != TrackState.Deleted && (key_prefix == null || p.Key.ToArray().Take(key_prefix.Length).SequenceEqual(key_prefix)))
+                    .Select(p =>
+                    (
+                        KeyBytes: p.Key.ToArray(),
+                        p.Key,
+                        p.Value.Item
+                    ))
+                    .OrderBy(p => p.KeyBytes, ByteArrayComparer.Default)
+                    .ToArray();
+            }
+            var uncached = FindInternal(key_prefix ?? new byte[0])
+                .Where(p => !dictionary.ContainsKey(p.Key))
+                .Select(p =>
+                (
+                    KeyBytes: p.Key.ToArray(),
+                    p.Key,
+                    p.Value
+                ));
+            using (var e1 = cached.GetEnumerator())
```

Comment:
I think this "intercalation" strategy should be done in a separate function, so it's clear and also testable (perhaps useful in other situations too).